src/main/java/org/springframework/integration/kafka/core/DefaultConnection.java

```diff
@@ -0,0 +1,312 @@
+			if (ErrorMapping.NoError() == errorCode) {
+				long[] offsets = offsetResponse.offsets(partition.getTopic(), partition.getId());
+				if (offsets.length == 0) {
+					throw new IllegalStateException("No error has been returned, but no offsets either");
+				}
+				resultBuilder.add(partition).withResult(offsets[0]);
+			}
+			else {
+				resultBuilder.add(partition).withError(errorCode);
+			}
+		}
+		return resultBuilder.build();
+	}
+
+	@Override
+	public Result<Void> commitOffsetsForConsumer(String consumerId, Map<Partition, Long> offsets) throws ConsumerException {
+		Map<TopicAndPartition, OffsetMetadataAndError> requestInfo =
+				MapIterate.collect(offsets, new CreateRequestInfoMapEntryFunction());
+		OffsetCommitResponse offsetCommitResponse = null;
+		try {
+			offsetCommitResponse = simpleConsumer.commitOffsets(
+					new OffsetCommitRequest(consumerId, requestInfo, kafka.api.OffsetCommitRequest.CurrentVersion(),
+							createCorrelationId(), simpleConsumer.clientId()));
+		}
+		catch (Exception e) {
+			throw new ConsumerException(e);
+		}
+		ResultBuilder<Void> resultBuilder = new ResultBuilder<Void>();
+		for (TopicAndPartition topicAndPartition : requestInfo.keySet()) {
+			if (offsetCommitResponse.errors().containsKey(topicAndPartition)) {
+				Partition partition = new Partition(topicAndPartition.topic(), topicAndPartition.partition());
+				resultBuilder.add(partition).withError((Short) offsetCommitResponse.errors().get(topicAndPartition));
+			}
+		}
+		return resultBuilder.build();
+	}
+
+	@Override
+	public Result<BrokerAddress> findLeaders(String... topics) throws ConsumerException {
+		TopicMetadataRequest topicMetadataRequest = new TopicMetadataRequest(Arrays.asList(topics), createCorrelationId());
+		TopicMetadataResponse topicMetadataResponse = null;
+		try {
+			topicMetadataResponse = simpleConsumer.send(topicMetadataRequest);
+		}
+		catch (Exception e) {
+			throw new ConsumerException(e);
+		}
+		ResultBuilder<BrokerAddress> resultBuilder = new ResultBuilder<BrokerAddress>();
+		for (TopicMetadata topicMetadata : topicMetadataResponse.topicsMetadata()) {
+			if (topicMetadata.errorCode() != ErrorMapping.NoError()) {
+				resultBuilder.add(new Partition(topicMetadata.topic(), -1)).withError(topicMetadata.errorCode());
+			}
+			else {
+				for (PartitionMetadata partitionMetadata : topicMetadata.partitionsMetadata()) {
+					Partition partition = new Partition(topicMetadata.topic(), partitionMetadata.partitionId());
+					if (ErrorMapping.NoError() == partitionMetadata.errorCode()) {
+						Broker leader = partitionMetadata.leader();
+						BrokerAddress result = new BrokerAddress(leader.host(), leader.port());
+						resultBuilder.add(partition).withResult(result);
+					}
+					else {
+						resultBuilder.add(partition).withError(partitionMetadata.errorCode());
+					}
+				}
+
+			}
+		}
+		return resultBuilder.build();
+	}
```

Comment:
Actually, that's not correct. The users of `KafkaOperations` shouldn't have any knowledge of the broker they need to talk to. The intent of this class is to wrap the low-level `SimpleConsumer` API.